src/Workspaces/VisualBasicTest/CodeGeneration/AddImportsTests.vb

```diff
@@ -0,0 +1,301 @@
+    Public Handler As System.EventHandler
+End Class</x>.Value,
+<x>Imports System
+Imports System.Collections
+Imports System.Collections.Generic
+Imports System.Diagnostics
+
+Class C
+    Public F As List(Of Integer)
+    Public Handler As EventHandler
+End Class</x>.Value)
+        End Sub
+
+        <Fact>
+        Public Sub TestImportNotAddedAgainIfAlreadyExists()
+            Test(
+<x>Imports System.Collections.Generic
+
+Class C
+    Public F As System.Collections.Generic.List(Of Integer)
+End Class</x>.Value,
+<x>Imports System.Collections.Generic
+
+Class C
+    Public F As System.Collections.Generic.List(Of Integer)
+End Class</x>.Value,
+<x>Imports System.Collections.Generic
+
+Class C
+    Public F As List(Of Integer)
+End Class</x>.Value)
+        End Sub
+
+        <Fact>
+        Public Sub TestUnusedAddedImportIsRemovedBySimplifier()
+            Test(
+<x>Class C
+    Public F As System.Int32
+End Class</x>.Value,
+<x>Imports System
+
+Class C
+    Public F As System.Int32
+End Class</x>.Value,
+<x>Class C
+    Public F As Integer
+End Class</x>.Value)
+        End Sub
+
+        <Fact>
+        Public Sub TestImportNotAddedIfGloballyImported()
+            Test(
+<x>Class C
+    Public F As System.Collections.Generic.List(Of Integer)
+End Class</x>.Value,
+<x>Class C
+    Public F As System.Collections.Generic.List(Of Integer)
+End Class</x>.Value,
+<x>Class C
+    Public F As List(Of Integer)
+End Class</x>.Value,
+globalImports:={"System.Collections.Generic"})
+
+        End Sub
+
+        <Fact>
+        Public Sub TestImportNotAddedForNamespaceDeclarations()
+            Test(
+<x>Namespace N
+End Namespace</x>.Value,
+<x>Namespace N
+End Namespace</x>.Value,
+<x>Namespace N
+End Namespace</x>.Value)
+        End Sub
+
+        <Fact>
+        Public Sub TestImportAddedAndRemovedForReferencesInsideNamespaceDeclarations()
+            Test(
+<x>Namespace N
+    Class C
+        Private _c As N.C
+    End Class
+End Namespace</x>.Value,
+<x>Imports N
+
+Namespace N
+    Class C
+        Private _c As N.C
+    End Class
+End Namespace</x>.Value,
+<x>Namespace N
+    Class C
+        Private _c As C
+    End Class
+End Namespace</x>.Value)
+        End Sub
+
+        <Fact>
+        Public Sub TestRemoveImportIfItMakesReferencesAmbiguous()
+            ' this Is not really an artifact of the AddImports feature, it is due
+            ' to Simplifier not reducing the namespace reference because it would 
+            ' become ambiguous, thus leaving an unused imports statement
+
+            Test(
+<x>Namespace N
+    Class C
+    End Class
+End Namespace
+
+Class C
+    Private F As N.C
+End Class
+</x>.Value,
+<x>Imports N
+
+Namespace N
+    Class C
+    End Class
+End Namespace
+
+Class C
+    Private F As N.C
+End Class
+</x>.Value,
+<x>Namespace N
+    Class C
+    End Class
+End Namespace
+
+Class C
+    Private F As N.C
+End Class
+</x>.Value)
+        End Sub
+
+    End Class
+End Namespace
```

Comment:
Are there VB-specific tests needed for partial namespaces? For example:

```VB
Imports System.Collections

Public Class C

	Public F1 As ArrayList
	Public F2 As System.Collections.Generic.List(Of Integer)

End Class
```

That code could reduce to this:

```VB
Imports System.Collections
Imports System.Collections.Generic

Public Class C

	Public F1 As ArrayList
	Public F2 As List(Of Integer)

End Class
```

or this:

```VB
Imports System.Collections

Public Class C

	Public F1 As ArrayList
	Public F2 As Generic.List(Of Integer)

End Class
```

Which one is more "VB"? Do we need an option to control that behavior?